reconcile/vault_replication.py

```diff
@@ -0,0 +1,392 @@
+import logging
+import re
+from collections.abc import Iterable
+from typing import (
+    Optional,
+    Union,
+    cast,
+)
+
+from reconcile.gql_definitions.jenkins_configs import jenkins_configs
+from reconcile.gql_definitions.jenkins_configs.jenkins_configs import (
+    JenkinsConfigsQueryData,
+    JenkinsConfigV1_JenkinsConfigV1,
+)
+from reconcile.gql_definitions.vault_instances import vault_instances
+from reconcile.gql_definitions.vault_instances.vault_instances import (
+    VaultInstanceV1_VaultReplicationConfigV1_VaultInstanceAuthV1,
+    VaultInstanceV1_VaultReplicationConfigV1_VaultInstanceAuthV1_VaultInstanceAuthApproleV1,
+    VaultReplicationConfigV1,
+    VaultReplicationConfigV1_VaultInstanceAuthV1,
+    VaultReplicationConfigV1_VaultInstanceAuthV1_VaultInstanceAuthApproleV1,
+    VaultReplicationJenkinsV1,
+)
+from reconcile.gql_definitions.vault_policies import vault_policies
+from reconcile.gql_definitions.vault_policies.vault_policies import (
+    VaultPoliciesQueryData,
+)
+from reconcile.utils import gql
+from reconcile.utils.vault import (
+    SecretAccessForbidden,
+    SecretNotFound,
+    SecretVersionNotFound,
+    VaultClient,
+    _VaultClient,
+)
+
+QONTRACT_INTEGRATION = "vault-replication"
+
+
+class VaultInvalidPaths(Exception):
+    pass
+
+
+class VaultInvalidAuthMethod(Exception):
+    pass
+
+
+def deep_copy_versions(
+    dry_run: bool,
+    source_vault: _VaultClient,
+    dest_vault: _VaultClient,
+    current_dest_version: int,
+    current_source_version: int,
+    path: str,
+) -> None:
+    for version in range(current_dest_version + 1, current_source_version + 1):
+        secret_dict = {"path": path, "version": version}
+
+        try:
+            secret, src_version = source_vault.read_all_with_version(secret_dict)
+            write_dict = {"path": path, "data": secret}
+            logging.info(["replicate_vault_secret", src_version, path])
+            if not dry_run:
+                dest_vault.write(
+                    secret=write_dict, decode_base64=False, force_write=True
+                )
+
+        except SecretNotFound:
+            # Handle the case where the version is already deleted from source, and
+            # we need to write a dummy version to the destination to keep the sync
+            # with versions in the source.
+            write_dummy_versions(
+                dry_run=dry_run,
+                dest_vault=dest_vault,
+                version_needed=version,
+                path=path,
+            )
```

Comment:
If I understand correctly, this is going to happen every time we have a source version > 1, correct? Can we simplify this? I mean, would it be simpler to just create the missing secrets whenever source version is > 1?

Reply:
this is not going to work like this. Looks better the way it is.